### ISSUE TITLE
setarch broken in docker packages from Debian stretch

### ISSUE THREAD

**Issue by juliusloman** [moby/moby#32839]
**Description**

setarch does not work in recent docker packages from Debian stretch repository. Arch and --uname-2.6 are being ignored in containers. 

Packages from Debian Jessie repository do not have this issue.

**Steps to reproduce the issue:**
1. Install docker-engine from debian strech repository 
`
deb https://apt.dockerproject.org/repo debian-stretch main
`
2. Run a base image with setarch command
`
docker run -t --rm debian setarch $(arch) --uname-2.6 uname -a
`
Issue is is not image specific, happens with other base images such as centos.

**Describe the results you received:**
Linux c49ed1228403 4.9.0-2-amd64 #1 SMP Debian 4.9.18-1 (2017-03-30) x86_64 GNU/Linux

**Describe the results you expected:**
Linux c49ed1228403 2.6.69-2-amd64 #1 SMP Debian 4.9.18-1 (2017-03-30) x86_64 x86_64 x86_64 GNU/Linux

**Additional information you deem important (e.g. issue happens only occasionally):**
It happens on the stretch version of Debian packages, packages from Jessie repository are OK.

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 17:50:07 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 17:50:07 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 123
Server Version: 17.04.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-2-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.56GiB
Name: lomo
ID: LR3Y:HX7L:T7ZI:RH6Z:QSLO:5E3Y:T5K3:4AQF:XMDH:VV5H:4WLO:WTN3
Docker Root Dir: /home/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://127.0.0.1:3128/
Https Proxy: http://127.0.0.1:3128/
No Proxy: localhost,127.0.0.1
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Host is running Debian Stretch.

**Comment by cpuguy83** [comment]
ping @justincormack

**Comment by ijc25** [comment]
> Packages from Debian Jessie repository do not have this issue.

Do you mean when the Jessie packages are installed on the same _Stretch_ based host system as the Stretch packages which exhibit the issue? Or do are you comparing with a Jessie based host system?

**Comment by juliusloman** [comment]
@ijc25 with Jessie packages installed on the same Stretch based host system. I've tried many different version of packages from both Stretch and Jessie (official) docker repositories. I'm experiencing this issue only on Stretch versions (host is Stretch based).

**Comment by ijc25** [comment]
Thanks.

One interesting (maybe?) observation is that it seems to work with `--privileged`:
```
[go:Moby] ijc@bokrug:moby$ docker run -t --rm debian setarch $(arch) --uname-2.6 uname -a
Linux 3bbd58b70877 4.9.0-2-amd64 #1 SMP Debian 4.9.18-1 (2017-03-30) x86_64 GNU/Linux
[go:Moby] ijc@bokrug:moby$ docker run --privileged -t --rm debian setarch $(arch) --uname-2.6 uname -a
Linux 8e29fce15ff8 2.6.69-2-amd64 #1 SMP Debian 4.9.18-1 (2017-03-30) x86_64 GNU/Linux
```
Perhaps it is seccomp or something which is preventing it working? (IIRC seccomp isn't enabled in the Jessie builds, so that would be consistent).

**Comment by ijc25** [comment]
`strace setarch $(arch) --uname-2.6 uname -a` when using `--privileged` contains:
```
personality(0x20000 /* PER_??? */)      = 0
```

whereas `strace setarch $(arch) --uname-2.6 uname -a` when using `--cap-add sys_ptrace` says:
```
personality(0x20000 /* PER_??? */)      = -1 EPERM (Operation not permitted)
```
which I suspect is the smoking gun. (aside: it's a bit annoying that `setarch` doesn't fail with a useful message and instead just blunders on)

**Comment by ijc25** [comment]
@juliusloman this should be fixed by #32965.
